File: LeetCode/src/dp/LongestIncreasingSubsequence.java
package dp;

import java.util.Arrays;

public class LongestIncreasingSubsequence {
    public int lengthOfLIS(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        int n = nums.length;
        int[] dp = new int[n]; // dp[i] stores the length of the longest increasing subsequence ending at index i
        Arrays.fill(dp, 1); // Initialize each element as a subsequence of length 1

        int maxLen = 1; // Initialize the maximum length to 1

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1); // Update the length of LIS ending at index i
                    maxLen = Math.max(maxLen, dp[i]); // Update the maximum length
                }
            }
        }

        return maxLen;

    }

    public static void main(String[] args) {
        LongestIncreasingSubsequence solution = new LongestIncreasingSubsequence();
        int[] nums = {10,9,2,5,3,7,101,18};
//        int[] nums = {4,10,4,3,8,9};
        System.out.println(solution.lengthOfLIS(nums)); // Output: 4
    }
}

File: LeetCode_Python/dp/lis/LongestIncreasingSubsequence.py
class LongestIncreasingSubsequence:
    def lengthOfLIS(self, nums: list[int]) -> int:
        n = len(nums)
        dp = [1] * n
        res = 1

        for i in range(1, n):
            for j in range(i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], dp[j] + 1)
                    res = max(res, dp[i])

        return res


if __name__ == "__main__":
    solution = LongestIncreasingSubsequence()

    print("Expected Output : 4")
    print("Actual Output : ", solution.lengthOfLIS([10, 9, 2, 5, 3, 7, 101, 18]))

    print("Expected Output : 4")
    print("Actual Output : ", solution.lengthOfLIS([0, 1, 0, 3, 2, 3]))

    print("Expected Output : 1")
    print("Actual Output : ", solution.lengthOfLIS([7, 7, 7, 7, 7, 7, 7]))
